Wait for the port to be open before opening the web browser.

tobb10001 opened the pull request:
Closes #412.

This PR adds functionality to check if a TCP port is open and to wait until a port opens, before opening a web browser. This functionality is used in `run.py`, in order to ensure that the UI is running before the browser opens.

The functionality is achieved using the `socket` and `threading` standard library modules. The port is checked to be open by trying to create a TCP connection. The waiting is made non-blocking by putting it into a separate thread.


I'm unsure about wether or not I found a good spot in the codebase to put my stuff. Also, this might be possible to do with async instead of a thread, but since `run()` is synchronous I don't think it fits here.

Tell me if I can improve anything.

rodja commented:
Works as expected. I've tested with

```py
from nicegui import app, ui

ui.label('Hello NiceGUI')
app.on_startup(lambda: time.sleep(5))

ui.run()
```

With different delay-lengths and also without the `app.on_startup`.

rodja commented on nicegui/helpers.py:
```diff
@@ -33,3 +37,29 @@ async def result_with_client():
+
+
+def schedule_browser(host: str, port: int) -> None:
+    # This function has to be non blocking. Therefore, we'll launch a thread, s.th.
```
This should be a docstring comment I think.

rodja commented on nicegui/helpers.py:
```diff
@@ -33,3 +37,29 @@ async def result_with_client():
+        webbrowser.open(f'http://{host}:{port}/')
+
+    host = host if host != "0.0.0.0" else "127.0.0.1"
+    threading.Thread(target=thread, args=(host, port), daemon=True).start()
```
Using a thread is a good choice here because the main event loop has not yet started.

rodja commented on nicegui/run.py:
```diff
@@ -88,7 +87,7 @@ def run(*,
 
     if show:
-        webbrowser.open(f'http://{host if host != "0.0.0.0" else "127.0.0.1"}:{port}/')
+        helpers.schedule_browser(host, port)
```
I like that you moved the code into helpers.py. This way we can use/adapt it for https://github.com/zauberzeug/nicegui/blob/main/nicegui/native_mode.py to fix #607.

rodja commented on nicegui/helpers.py:
```diff
@@ -33,3 +37,29 @@ async def result_with_client():
+
+
+def port_open_tcp(host: str, port: int) -> bool:
+    # Check if the port is open by checking if a TCP connection can be established.
```
Should be a docstring

rodja commented on nicegui/helpers.py:
```diff
@@ -33,3 +37,29 @@ async def result_with_client():
         globals.handle_exception(e)
+
+
+def port_open_tcp(host: str, port: int) -> bool:
```
Could you rename this to something like `is_port_open`?

tobb10001 commented:
Thanks for the feedback! I've implemented it.

I've also gotten some ideas on how to test the stuff I added, so I implemented them as well.

Speaking of tests, I've found that I have some trouble with them on my machine, since some of them were failing (on `main`), although they didn't fail in CI on `main`.

I've got some of them to work by installing `pytest-asyncio`, which I did because before I got the warning that the option `asyncio_mode` (defined in `pytest.ini`) is unknown. If this is common, maybe it should be put into `tests/requirements.txt`.

But the tests in `tests/test_scene.py` still aren't working on my machine.


I can provide detail if you want to, but I don't think it's a big deal. I just wanted to report this here, because the contributing guidelines state that the tests should pass, but since they also don't pass on `main` on my machine, I don't think it's got something to do with my proposed changes.

rodja commented:
> I've got some of them to work by installing pytest-asyncio, which I did because before I got the warning that the option asyncio_mode (defined in pytest.ini) is unknown. If this is common, maybe it should be put into tests/requirements.txt.

Thanks. We missed that dependency. I've added it straight away.

> But the tests in tests/test_scene.py still aren't working on my machine. I can provide detail if you want to, but I don't think it's a big deal. I just wanted to report this here, because the contributing guidelines state that the tests should pass, but since they also don't pass on main on my machine, I don't think it's got something to do with my proposed changes.

Could you create a separate issue for this. I would like to investigate this further. For now its ok you ignore the passing of test_scene.py, interference with this PR is not very likely.

falkoschindler commented:
The scene tests have been red on my machine as well for quite a while now. I also ignored it since they are succeeding on GitHub. It would definitely be good to track and fix this issue.

rodja commented:
@falkoschindler indeed, the tests two new tests are failing for me too:

```
$ pytest -k helpers
/opt/homebrew/lib/python3.11/site-packages/pytest_asyncio/plugin.py:173: DeprecationWarning: You're using an outdated version of pytest. Newer releases of pytest-asyncio will not be compatible with this pytest version. Please update pytest to version 7 or later.
  warnings.warn(
================================================================= test session starts ==================================================================
platform darwin -- Python 3.11.1, pytest-6.2.5, py-1.11.0, pluggy-1.0.0
driver: Chrome
sensitiveurl: .*
rootdir: /Users/rodja/Projects/nicegui, configfile: pytest.ini
plugins: variables-2.0.0, xdist-3.2.0, asyncio-0.20.3, html-3.2.0, selenium-4.0.0, flaky-3.7.0, flakefinder-1.1.0, base-url-2.0.0, metadata-2.0.4, anyio-3.6.2
asyncio: mode=Mode.AUTO
collected 115 items / 112 deselected / 3 selected                                                                                                      

tests/test_helpers.py F.F                                                                                                                        [100%]

======================================================================= FAILURES =======================================================================
__________________________________________________________________ test_is_port_open ___________________________________________________________________

    def test_is_port_open():
        with contextlib.closing(socket.socket(socket.AF_INET, socket.SOCK_STREAM)) as sock:
    
            sock.bind(('127.0.0.1', 0))  # port = 0 => the OS chooses a port for us
            host, port = sock.getsockname()
    
            # port bound, but not opened yet
>           assert not helpers.is_port_open(host, port)

tests/test_helpers.py:16: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

host = '127.0.0.1', port = 55208

    def is_port_open(host: str, port: int) -> bool:
        """Check if the port is open by checking if a TCP connection can be established."""
        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        try:
>           sock.connect((host, port))
E           TimeoutError: [Errno 60] Operation timed out

nicegui/helpers.py:46: TimeoutError
_____________________________________________________________ test_schedule_browser_cancel _____________________________________________________________

monkeypatch = <_pytest.monkeypatch.MonkeyPatch object at 0x11ec3f650>

    def test_schedule_browser_cancel(monkeypatch):
    
        called_with_url = None
    
        def mock_webbrowser_open(url):
            nonlocal called_with_url
            called_with_url = url
    
        monkeypatch.setattr(webbrowser, "open", mock_webbrowser_open)
    
        # This test doesn't need to open a port, but it binds a socket, s.th. we can be sure
        # it is NOT open.
    
        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        sock.bind(('127.0.0.1', 0))
        host, port = sock.getsockname()
    
        thread, cancel_event = helpers.schedule_browser(host, port)
    
        cancel_event.set()
    
        time.sleep(0.2)
    
>       assert not thread.is_alive()
E       assert not True
E        +  where True = <bound method Thread.is_alive of <Thread(Thread-2 (in_thread), started daemon 6170193920)>>()
E        +    where <bound method Thread.is_alive of <Thread(Thread-2 (in_thread), started daemon 6170193920)>> = <Thread(Thread-2 (in_thread), started daemon 6170193920)>.is_alive

tests/test_helpers.py:75: AssertionError
=============================================================== short test summary info ================================================================
FAILED tests/test_helpers.py::test_is_port_open - TimeoutError: [Errno 60] Operation timed out
FAILED tests/test_helpers.py::test_schedule_browser_cancel - assert not True
```

falkoschindler commented:
Regarding test_scene.py, I opened a new issue: #643

tobb10001 commented:
The first test `test_is_port_open` is an OS-specific problem. My machine actively refuses the connection, resulting in the `ConnectionRefusedError`, others don't send this, s.th. nothing happens, resulting in a `TimeoutError`.

I will adjust the system to handle both cases.

For the second test: Could you test the behaviour on your machine, if you set the timeout [here](https://github.com/tobb10001/nicegui/blob/792eaf87968688a23ffd2047f5161396e5fe435b/tests/test_helpers.py#L73) to like 0.5 seconds? Currently that's my only guess, that I set it too low.

tobb10001 commented:
I rebased on `main` earlier today and another test `tests/test_auto_context.py::test_autoupdate_on_async_timer_callback` fails now as well, but again, I observe the same behaviour on `main` on my machine...

Should I open another issue?

rodja commented:
@tobb10001 I found that just binding a socket (without listening) took very long. While `test_is_port_open` succeeded after about 10 sec, the cancelling test failed. I modified both so they pass on Mac. Please have a look and try on your side...

rodja commented on tests/test_helpers.py:
```diff
@@ -0,0 +1,74 @@
+
+        try:
+            # port bound, but not opened yet
+            assert called_with_url is None
```
instead of comments I find it often better to provide a assert message like

```py
assert called_with_url is None, 'should not have been called because port is only bound, not opened'

rodja commented:
@tobb10001 We would really like this to be part of the upcoming 1.2.3 release. With this code we can probably also fix #607. As soon as you have tried and reviewed my changes we are ready to merge.

tobb10001 commented:
@rodja Sounds great! 👍🏽

I'll probably get it done tomorrow, maybe even today.

tobb10001 commented on tests/test_helpers.py:
```diff
@@ -0,0 +1,74 @@
+from nicegui import helpers
+
+
+def test_is_port_open():
+    with contextlib.closing(socket.socket(socket.AF_INET, socket.SOCK_STREAM)) as sock:
+        sock.bind(('127.0.0.1', 0))  # port = 0 => the OS chooses a port for us
+        sock.listen(1)
+        host, port = sock.getsockname()
+    assert not helpers.is_port_open(host, port), 'after closing the socket, the port should be free'
+
+    with contextlib.closing(socket.socket(socket.AF_INET, socket.SOCK_STREAM)) as sock:
+        sock.bind(('127.0.0.1', port))
+        sock.listen(1)
+        assert helpers.is_port_open(host, port), 'after opening the socket, the port should be detected'
```
Now that I look at this version again, I would actually split this into two different test cases. But I will leave this up to you as maintainers.

```suggestion
def test_is_port_open_when_open():
    with contextlib.closing(socket.socket(socket.AF_INET, socket.SOCK_STREAM)) as sock:
        sock.bind(('127.0.0.1', 0))  # port = 0 => the OS chooses a port for us
        sock.listen(1)
        host, port = sock.getsockname()
    assert not helpers.is_port_open(host, port), 'after closing the socket, the port should be free'


def test_is_port_open_when_closed():
    with contextlib.closing(socket.socket(socket.AF_INET, socket.SOCK_STREAM)) as sock:
        sock.bind(('127.0.0.1', port))
        sock.listen(1)
        assert helpers.is_port_open(host, port), 'after opening the socket, the port should be detected'
```

tobb10001 commented on tests/test_helpers.py:
```diff
@@ -0,0 +1,74 @@
+    # find a free port ...
+    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+    sock.bind(('127.0.0.1', 0))
+    sock.listen(1)
```
I originally left this out intentionally, in order to have the port allocated.

I think in this configuration it might theoretically be possible for another process to open the port in the meantime. However, I assume this is never going to happen in practice, so I'm fine with it.

tobb10001 reviewed (commented):
Looks good to me. :+1:
